firedrake/variational_solver.py

```diff
@@ -83,6 +86,13 @@ def __init__(self, F, u, bcs=None, J=None,
         self.form_compiler_parameters = form_compiler_parameters
         self._constant_jacobian = False
 
+        # used in NonlinearVariationalSolverMixin._ad_annotate_solve
+        self._ad_F = F
+        self._ad_u = u
+        self._ad_bcs = bcs
+        self._ad_J = J
+        self._ad_kwargs = {'Jp': Jp, 'form_compiler_parameters': form_compiler_parameters, 'is_linear': is_linear}
```

Comment:
Since we now control the mixin class, can we not just use the properties that already exist? This sets up a contract between mixin of a related class to the `NLVP` that kind of smells bad.

Reply:
Or could we just create `NonlinearVariationalProblemMixin` that only for these lines?

Reply:
fixed. I just created `NonlinearVariationalProblemMixin` for now to be consistent with what we do for the other classes.